src/mon/OSDMonitor.cc

```diff
@@ -2061,8 +2061,37 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
     }
   }
 
+  auto m = op->get_req<MOSDBoot>();
+  int from = m->get_orig_source_inst().name.num();
   // health
   health_check_map_t next;
+
+  if (auto type_iter = m->metadata.find("osd_type");
+      type_iter != m->metadata.end()) {
+    const auto &otype = type_iter->second;
+    // m->metadata["osd_type"] must be "crimson", classic doesn't send osd_type
+    if (otype == "crimson" && !tmp.get_allow_crimson()) {        
```

Comment:
We handle `MOSDBoot` messages in `OSDMonitor::preprocess_boot`, we should probably try to reuse this logic and set the warning from there. As mentioned, we shouldn't handle these messages in `encode_pending`.

Getting a health warning from `preprocess_boot` would be tricky with the current warning design we have. It seems that we encode health checks only from `encode_pending`. Optimally, we would have been able to add the health check in `OSDMap::check_health`. However, OSDMap current structure is not able to know about our warning.
We could  try adding the warning information to OSDMap (similiarly to what we did with `OSDMap::allow_crimson`), though I don't think that it's worth the overhead of storing that in every OSDMap shared.

If a health warning would end up being too tricky to implement, we could perhaps reply to the OSD (MOSDBootReply?) so that the OSD will go down once with a proper abort message once this message is received?
@athanatos, do you have a better suggestion?